src/Algebra/Graph/Labelled/AdjacencyMap/Algorithm/Dijkstra.hs

```diff
@@ -0,0 +1,106 @@
+module Algebra.Graph.Labelled.AdjacencyMap.Algorithm.Dijkstra where
+
+import Algebra.Graph.Label (Capacity, Distance, Semiring(..), (<+>), zero)
+import Algebra.Graph.Labelled.AdjacencyMap.Internal (AdjacencyMap(..))
+import Control.Monad (sequence_, unless, when)
+import Control.Monad.Trans.State.Strict (State, execState, get, gets, put)
+import Data.Heap (FstMaxPolicy, FstMinPolicy, Heap, HeapItem(..))
+import qualified Data.Heap as Heap
+import Data.Map (Map, (!))
+import qualified Data.Map.Strict as Map
+
+type DijkstraForDistance a b = DijkstraState FstMinPolicy a (Distance b)
+
+type DijkstraForCapacity a b = DijkstraState FstMaxPolicy a (Capacity b)
+
+-- | /O(|V| + |E| * log|V|) `dijkstra` is the main function using
+-- both `initialize` and `exploreVertices`.
+dijkstra ::
+     (Eq e, Ord a, HeapItem p (e, a), Semiring e)
+  => a
+  -> AdjacencyMap e a
+  -> DijkstraState p a e
+dijkstra a am =
+  flip execState (DijkstraState Heap.empty Map.empty Map.empty Map.empty) $ do
+    initialize a am
+    exploreVertices am
+
+data DijkstraState p a e = DijkstraState
+  { heap :: Heap p (e, a)
+  , explored :: Map a Bool
```

Comment:
This is only used for further optimization. The time complexity would remain the same even if this was removed.

Reply:
See my comment below about avoiding premature optimisation. I would start with the simplest implementation first, because it's easier to implement correctly, easier to review, easier to integrate into the library, etc.